xmake/rules/vala/xmake.lua

```diff
@@ -58,67 +58,82 @@ rule("vala.build")
+                for _, dep in ipairs(target:orderdeps()) do
+                    if dep:is_shared() or dep:is_static() then
+                        local vapifile = dep:data("vala.vapifile")
+                        if vapifile then
+                            table.join2(argv, path(vapifile))
+                        end
                     end
                 end
+            else
+                local vapifile = target:data("vala.vapifile")
+                if vapifile then
+                    table.insert(argv, path(vapifile, function (p) return "--vapi=" .. p end))
+                end
+                local headerfile = target:data("vala.headerfile")
+                if headerfile then
+                    table.insert(argv, "-H")
+                    table.insert(argv, path(headerfile))
+                end
             end
-        else
-            local vapifile = target:data("vala.vapifile")
-            if vapifile then
-                table.insert(argv, path(vapifile, function (p) return "--vapi=" .. p end))
+            local vapidir = target:data("vala.vapidir")
+            if vapidir then
+                table.insert(argv, path(vapidir, function (p) return "--vapidir=" .. p end))
             end
-            local headerfile = target:data("vala.headerfile")
-            if headerfile then
-                table.insert(argv, "-H")
-                table.insert(argv, path(headerfile))
+            local valaflags = target:data("vala.flags")
+            if valaflags then
+                table.join2(argv, valaflags)
             end
+            table.insert(argv, path(sourcefile_vala))
         end
-        local vapidir = target:data("vala.vapidir")
-        if vapidir then
-            table.insert(argv, path(vapidir, function (p) return "--vapidir=" .. p end))
-        end
-        local valaflags = target:data("vala.flags")
-        if valaflags then
-            table.join2(argv, valaflags)
-        end
-        table.insert(argv, path(sourcefile_vala))
+
+        batchcmds:show_progress(opt.progress, "${color.build.object}compiling.vala")
         batchcmds:vrunv(valac.program, argv)
-        batchcmds:compile(sourcefile_c, objectfile)
+
+        for _, sourcefile_c in ipairs(sourcefiles_c) do
+            batchcmds:show_progress(opt.progress, "${color.build.object}compiling.c %s", sourcefile_c)
+            batchcmds:compile(sourcefile_c, target:objectfile(sourcefile_c))
+        end
 
         -- add deps
-        batchcmds:add_depfiles(sourcefile_vala)
-        batchcmds:set_depmtime(os.mtime(objectfile))
-        batchcmds:set_depcache(target:dependfile(objectfile))
+        for _, sourcefile_vala in ipairs(sourcefiles_vala.sourcefiles) do
+            batchcmds:add_depfiles(sourcefile_vala)
+        end
+        for _, objectfile in ipairs(target:objectfiles()) do
+            batchcmds:set_depmtime(os.mtime(objectfile))
+            batchcmds:set_depcache(target:dependfile(objectfile))
```

Comment:
This will break incremental compilation.

Reply:
Is there a way to set batchcmd depend from hash instead of modify time if to move compilation of **only** outputed C code into before_buildcmd_file?

Reply:
Doesn't matter anymore, see latest commit